Add frame matrix

LaDilettante opened the pull request:
I extract the code that parses tribble(...) arguments into `extract_data()`, which returns `list(frame_names, frame_rest)`. Then I re-use `extract_data()` in both `tribble()` and `frame_matrix()`. I'd appreciate thoughts on this.

If this strategy / function names seem okay, I'll submit more PR for edge cases.

krlmlr commented on R/tribble.R:
```diff
@@ -28,16 +28,56 @@
+  frame_mat
+}
+
+extract_data <- function(...) {
```
The function name isn't very meaningful.

krlmlr commented on R/tribble.R:
```diff
@@ -28,16 +28,56 @@
-      out <- rep(list(logical()), length(frame_names))
-      names(out) <- frame_names
-      return(as_tibble(out))
+      return(list(frame_names = frame_names, frame_rest = NULL))
```
Are the corner cases covered by the tests?

krlmlr commented on R/tribble.R:
```diff
@@ -28,16 +28,56 @@
   # Extract the names.
   frame_names <- character()
   i <- 1
   while (TRUE) {
```
Would you mind extracting the loop to a function?

krlmlr commented on R/tribble.R:
```diff
@@ -82,19 +122,7 @@ tribble <- function(...) {
-  # Create a tbl_df and return it
-  names(frame_col) <- frame_names
-  as_tibble(frame_col)
+  list(frame_names = frame_names, frame_rest = frame_rest)
```
The error checking and the composition of the returned list could also form a separate function.

krlmlr commented:
Oddly, GitHub includes my already committed changes to master in the diff. Could you please rebase against master?

LaDilettante commented:
Did you write some code for this yourselves? I pull rebase from master and saw some new code for edge cases. Just curious.

I think I've addressed all of your comments in the last PR (thanks as always!). Is there anything more I need to do to get Github's new "Requested changes" to go away?

krlmlr commented on tests/testthat/test-tribble.R:
```diff
@@ -104,3 +126,48 @@ test_that("tribble recognizes quoted non-formula call", {
+  expect_equal(result, expected)
+})
+
+test_that("frame_matrix creates 0x0 matrix with column names when
```
Please double-check the test title.

krlmlr replied:
Is it really a 0x0 matrix that's created here?

krlmlr commented on R/tribble.R:
```diff
@@ -27,17 +27,80 @@
+  frame_names <- extract_frame_names_from_dots(dots)
+
+  # Extract the data
+  if (length(frame_names) == length(dots)) {
```
Is this special case really necessary? The code might look cleaner if you just did `frame_rest <- dots[-seq_along(frame_names)]` .

krlmlr commented on R/tribble.R:
```diff
@@ -27,17 +27,80 @@
+extract_frame_data_from_dots <- function(...) {
   dots <- list(...)
 
+  if (length(dots) == 0) {
```
This shortcut might become unnecessary, too.

krlmlr commented on R/tribble.R:
```diff
@@ -27,17 +27,80 @@
 tribble <- function(...) {
+  data <- extract_frame_data_from_dots(...)
 
+  if (is.null(data$frame_rest)) {
```
Safer to check `length() == 0` here.

krlmlr commented:
I have no idea why GitHub is doing the things it does. Anyway, if you just do `git rebase origin/master; git push -u LaDilettante HEAD --force` on your branch (assuming `origin` is the upstream remote and `LaDilettante` is yours) the spurious changes should go away.

LaDilettante commented:
My apology for being late with my follow-up -- I've been busying applying for jobs.

I'm committed to close this issue soon. Let me know if it needs more work.

krlmlr commented:
Thanks. Could you please rebase your branch against current master, and force-push? GitHub seems to have a tough time isolating the changes from your PR from those that are already in master.

LaDilettante commented:
@krlmlr Is this rebasing sufficient to your need?

krlmlr commented on R/tribble.R:
```diff
@@ -27,19 +27,75 @@
+
+  frame_mat <- matrix(data$frame_rest, ncol = length(data$frame_names), byrow = TRUE)
+
+  frame_col <- lapply(seq_len(ncol(frame_mat)), function(i) {
```
Could you please extract everything after `extract_frame_data_from_dots(...)` to a separate function, and perhaps also `function(i) {...}` ?

LaDilettante replied:
I could, but would appreciate more pointers about why. The implementation after `extract_frame_data_from_dots()` is very unique, so if we extract everything after it into a function, this function won't be reused anywhere but here. Note that the implementation of `tribble` and `frame_matrix` are quite different after `extract_frame_data_from_dots(...)`, so I don't think any reuse is possible.

krlmlr replied:
It's based on the principles of "clean code" I learned from Robert C. Martin's book (not comprehensive):

- [x] Each function should do one thing (and do it well)
- [x] The purpose of the function should be obvious from its name
- [ ] Prefer short pure functions with few arguments
- [ ] All code in a function should operate on the same level of abstraction, so that the logic of each function is easy to understand

In addition, by extracting a function you could write a unit test specifically for that function, but that's currently not top priority.

krlmlr commented on R/tribble.R:
```diff
@@ -27,19 +27,75 @@
+frame_matrix <- function(...) {
+  data <- extract_frame_data_from_dots(...)
+
+  if (any(vapply(data$frame_rest, needs_list_col, logical(1)))) {
```
Same here -- please extract everything after `extract_frame_data_from_dots(...)` to a function.

LaDilettante commented on R/tribble.R:
```diff
@@ -27,19 +27,75 @@
 tribble <- function(...) {
+  data <- extract_frame_data_from_dots(...)
 
+  if (length(data$frame_rest) == 0) {
```
Later in `extract_frame_data_from_dots()`, `frame_rest` is elegantly returned as a list with `frame_rest <- dots[-seq_along(frame_names)]`. In the empty case, this returns `frame_rest` as `list()`. I think it's better to keep `frame_list` consistently as a list no matter empty or not. 

If we remove the special case here, we have to add a special case in `extract_frame_data_from_dots()` so that `frame_rest` is `logical()`.

krlmlr replied:
But you could do here:

```r
if (length(data$frame_rest) == 0) {
  data$frame_rest <- logical()
}
```

I prefer to avoid shortcuts for edge cases if at all possible, because they are easy to overlook if the code needs to be edited.

LaDilettante commented:
@krlmlr I add a PR for the smaller stuff. 

If I extract out the code after `extract_frame_data_from_dots(...)`, those codes won't be reused anywhere else. I'm thus unsure about why we should do that. Could you explain a bit more about what your thinking behind requesting that extraction so that I can name the function / return appropriately?

LaDilettante commented:
@krlmlr Everything extracted out as requested.

LaDilettante commented:
@krlmlr I moved things around as requested -- thanks for being patient with your guidance.

Two more things we may want to do
- Should we use `frame_matrix()` while `frame_data()` will be phased out? Would `mibble()` be too obscure?
- I realize that we should also allow `frame_matrix()` to take in row names. If you think it's a good idea, I can file an issue and pick this up in a month.

krlmlr commented:
Thanks!

@hadley: Can you suggest an alternative name for `frame_matrix()` ? Are we going to keep the `frame_data()` alias?
